### PR TITLE
Ensure Distribution.sample() result is detached

### PR DESCRIPTION
**Pull request by fritzo**
This fixes a bug in `TransformedDistribution` whereby `.sample()` was accidentally reparameterized. It also fixes a couple other minor `.sample()` bugs and adds two tests to distinguish the behavior of `.sample()` and `.rsample()`.

Previous review by @alicanb at https://github.com/probtorch/pytorch/pull/125

### CODE REVIEW

**Comment by fritzo** [comment]
cc @apaszke

File: torch/distributions/bernoulli.py [hunk 1]
```diff
@@ -72,7 +72,8 @@ def param_shape(self):
 
     def sample(self, sample_shape=torch.Size()):
         shape = self._extended_shape(sample_shape)
-        return torch.bernoulli(self.probs.expand(shape))
+        x = torch.bernoulli(self.probs.expand(shape))
```
Comment (colesbury) [inline torch/distributions/bernoulli.py]: Use:

```
with torch.no_grad():
  return torch.bernoulli(self.probs.expand(shape))
```

instead of detaching at the end

File: test/test_distributions.py [hunk 2]
```diff
@@ -364,6 +364,32 @@ def _check_enumerate_support(self, dist, examples):
...
+                sample = dist.sample()
+                self.assertFalse(sample.requires_grad,
+                                 msg='{} example {}/{}, .sample() is not detached'.format(
+                                     Dist.__name__, i + 1, len(params)))
+
+    def test_rsample_requires_grad(self):
+        for Dist, params in EXAMPLES:
+            for i, param in enumerate(params):
+                variable_params = [p for p in param.values() if getattr(p, 'requires_grad', False)]
+                if not variable_params:
```
Comment (apaszke) [inline test/test_distributions.py]: a nicer way to do the same would be to
```python
if not any(getattr(p, 'requires_grad', False) for p in param.values()):
    continue
```

File: torch/distributions/bernoulli.py (hunk 1 above)
Comment (apaszke) [reply torch/distributions/bernoulli.py]: This will also make it faster, because `hasattr` is quite expensive

Comment (fritzo) [reply torch/distributions/bernoulli.py]: Thanks, I'll start using `torch.no_grad()`.

**Comment by apaszke** [comment]
@fritzo all contbuilds are failing

File: torch/distributions/distribution.py [hunk 3]
```diff
@@ -72,8 +72,9 @@ def sample(self, sample_shape=torch.Size()):
         Generates a sample_shape shaped sample or sample_shape shaped batch of
         samples if the distribution parameters are batched.
         """
-        z = self.rsample(sample_shape)
-        return z.detach() if hasattr(z, 'detach') else z
+        with torch.no_grad():
+            x = self.rsample(sample_shape)
+        return x.detach() if hasattr(x, 'detach') else x
```
Comment (fritzo) [inline torch/distributions/distribution.py]: I tried to remove this `.detach()` but some `.sample()`s still ended up requiring grad.

Comment (apaszke) [reply torch/distributions/distribution.py]: Do you know which distributions had this problem? `x` should never require grad in here

Comment (fritzo) [reply torch/distributions/distribution.py]: This fails for at least `Dirichlet` and `Beta`:
```py
param = Variable(torch.ones(2), requires_grad=True)
dist = torch.distributions.Dirichlet(param)
with torch.no_grad():
    dist.rsample().requires_grad
# True
```

Comment (apaszke) [reply torch/distributions/distribution.py]: Looks like Python functions don't respect `GradMode`. Will send a fix today

Comment (apaszke) [reply torch/distributions/distribution.py]: See #5184

Comment (fritzo) [reply torch/distributions/distribution.py]: Thanks, that fixed it!

File: torch/distributions/transformed_distribution.py [hunk 4]
```diff
@@ -53,10 +53,11 @@ def sample(self, sample_shape=torch.Size()):
...
         """
-        x = self.base_dist.sample(sample_shape)
-        for transform in self.transforms:
-            x = transform(x)
-        return x
+        with torch.no_grad():
+            x = self.base_dist.sample(sample_shape)
+            for transform in self.transforms:
+                x = transform(x)
+            return x
```
Comment (apaszke) [inline torch/distributions/transformed_distribution.py]: isn't this just `self.rsample(sample_shape)`, but in context of `no_grad`?

**Review by apaszke — APPROVED** [review]
